File: backend/dataset/maxPathSum/maxPathSum_10.py
import sys
class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        self.answer =  - sys.maxsize - 1
        def dfs(node):
            if not node:
                return 0
            left_sub_tree = dfs(node.left)
            right_sub_tree = dfs(node.right)
            self.answer = max(self.answer , left_sub_tree + right_sub_tree + node.val)
            return max(0 , node.val + max(left_sub_tree , right_sub_tree))
        dfs(root)
        return self.answer

File: backend/CDataset/maxPathSum/maxPathSum_10.cpp
class Solution {
public:
    class TreeNode{
        public:
        TreeNode* left;
        TreeNode* right;
        int val;
    };
  int maxPathSum(TreeNode* root) {
    int maximumPath = INT_MIN;
    auto updateMax = [&maximumPath](int num){
      if(maximumPath < num) maximumPath = num;
    };
    function<int(TreeNode* node)> recurser;
    recurser = [&recurser, &updateMax](TreeNode* node){
      if(!node) return 0;
      updateMax(node->val);
      int leftPath = recurser(node->left);
      int rightPath = recurser(node->right);
      int longerPath = max(leftPath, rightPath);
      if(leftPath>0 && rightPath>0) updateMax(node->val + leftPath + rightPath);
      else updateMax(node->val + longerPath);
      return max(0, node->val + longerPath);
    };
    recurser(root);
    return maximumPath;
  }
};
